### DISCUSSION TITLE
how to make a nuscript standalone executable?

### DISCUSSION THREAD

**Discussion by jakara** [nushell/nushell#16663] [Q&A]
### Question

like pyinstaller for python, is there a tool or method I can package a nuscript file into a standalone executable, that runs on OS without nushell installed?

**Comment by hgkamath** [comment]
Just a user here:  
Python interpreter can be embedded into rust (pyO3, rustpython) , javascript (brython) , java (Jython). The use case is that python acts like a plugin-language or DSL within the context of a larger application. 

On the surface of it, what you ask is a very unlikely direction, as shells are meant to run human readable interactive scripts.
Shell scripts run on the outside (before) invoked programs, ergo, the name 'shell'. They do tooling, do only simple things and invoke other big programs to do the heavy work. Shell script interpreters, interpret script-lines string-by-string, without doing an entire compilation to an optimized binary machine-code.
see https://www.nushell.sh/book/how_nushell_code_gets_run.html   
If the necessary goal was to make a standalone executable, then one must choose a compilable-programing-language. There are only a few use cases where scripts are run inside standalone binaries, like installers (NSIS) and self-extractors (SFX). 

pyinstaller, that you mentioned is like a SFX bundler, uses the egg distribution format. It puts all the dependencies into a single folder, separating the distribution of dependencies (myscript.zip) from the bootloaded-script itself (myscript.exe). 
https://pyinstaller.org/en/stable/operating-mode.html
As it bundles the interpreter as well, it has the liberty after unpacking to invoke the interpreter, both while installing or later.
The bootloader part of pyinstaller is written in ```C```. https://github.com/pyinstaller/pyinstaller  
In any case, pyinstaller is an external 3rd party tool to python itself.

It seems to me, what you're asking is a way to embed the nushell nushell-script-interpreter (or byte-code-interpreter) in a C/Rust program, embed a static data-string containing the script file (or nushell bytecode) as input, The main function calls the ```eval``` function and supplies it with the input-data-string. Then statically compile all the dependencies into a single executable. But, usually, it doesn't make sense to do this, as now, the nushell library-interpreter essentially becomes a one-trick-pony (unlike the use-case of embedded python). Why make a one-trick pony, when you can have the whole multipurpose shell? You don't gain any speed/performance benefits as it is the same interpreter, but have the downside of being large/bloated.

So, it depends on whether nushell has been developed in such a way, so has to have a statically linkable ```libnushell.a``` library to export the interpreter module, which in-turn can be imported/included while statically compiling.

A nushell developer can comment if such library exists.

Perhaps, in time, someone will create a 3rd party nushell sfx bundler like pyinstaller.

**Comment by rgwood** [comment]
Nu does not have a way to do this right now. It's something we'd like to have eventually.

I'd lean toward taking the same approach as `deno compile`, where we embed the script inside a Nu executable: <https://github.com/denoland/deno/pull/8539> <https://github.com/denoland/deno/pull/9141>

```rust
/// This function will try to run this binary as a standalone binary
/// produced by `deno compile`. It determines if this is a stanalone
/// binary by checking for the magic trailer string `D3N0` at EOF-12.
/// After the magic trailer is a u64 pointer to the start of the JS
/// file embedded in the binary. This file is read, and run. If no
/// magic trailer is present, this function exits with Ok(()).
```
> `--target` allows to cross-compile binary to different target architectures by fetching appropriate binary from remote server on first run.

Related discussion on Discord: https://discordapp.com/channels/601130461678272522/615962413203718156/1077740195618758688

**Comment by jakara** [comment]
> what you're asking is a way to embed the nushell nushell-script-interpreter (or byte-code-interpreter) in a C/Rust program,
@hgkamath 


no, I don't want to embed nushell in some seriously compiled(C/rust) programs, what I'm thinking is to make nu scripts more "portable",  so it can run on some bared OS, in bash maybe,  and it could work like some native tools, curl, eg.
I see nu language designed with a lot reliable, efficient features, such a waste to use as just a "shell", IMHO 😝

**Comment by rgwood** [comment]
I’m not sure I understand what you’re looking for. I agree that the Nu language is powerful and I’ve proposed one way that you could compile Nu scripts into standalone executables. Could you clarify what makes the  Pyinstaller approach different?

**Comment by jakara** [comment]
> I’m not sure I understand what you’re looking for. I agree that the Nu language is powerful and I’ve proposed one way that you could compile Nu scripts into standalone executables. Could you clarify what makes the Pyinstaller approach different?

Forget Pyinstaller, what `deno compile` did is exactly what I want from Nu

**Comment by salsa-dev** [comment]
@jakara having standalone script for execution could be achieved by using env in your script
```
#!/usr/bin/env nu
```
it is described in docs
https://www.nushell.sh/book/scripts.html#shebangs

this is probably the most you can achieve

**Comment by WindSoilder** [comment]
Closing the issue because the question is answered, currently you can use shebang to achieve something like this.

**Comment by XiNiHa** [comment]
This issue is about creating executables that can be run  **without having nushell installed,** so shebang can't be the answer. Reopen the issue please.

**Comment by fdncred** [comment]
Right now, the only way to do this is to build an executable with that script embedded similar to this https://github.com/sophiajt/nu_app or maybe something like this https://github.com/cablehead/how-to-embed-nu
